components/crud-web-apps/jupyter/backend/apps/common/form.py

```diff
@@ -163,6 +163,22 @@ def set_notebook_cpu(notebook, body, defaults):
     container = notebook["spec"]["template"]["spec"]["containers"][0]
 
     cpu = get_form_value(body, defaults, "cpu")
+    cpuLimit = get_form_value(body, defaults, "cpuLimit")
+
+    limit_factor = utils.load_spawner_ui_config()["cpu"].get("limitFactor")
+    if not cpuLimit and limit_factor != "none":
+        cpuLimit = str(round((float(cpu) * float(limit_factor)), 1))
+        if not cpu:
```

Comment:
I'm pretty sure I needed to add this for something. I'll investigate.

Reply:
I think this was needed for the situation where `cpu` is set to readOnly and the limitFactor is set. In this situation the limit needs to be calculated from the value in the `spawner_ui_config.yaml` as no cpu value is included in the request.

Reply:
Then we can still remove the check. The `cpu = get_form_value(body, defaults, "cpu")` line above ensures that you will have a `cpu` value. The helper function [get_form_value](https://github.com/kubeflow/kubeflow/blob/43e03d4ae176c735e53843f8f0b3f7d7ed79756b/components/crud-web-apps/jupyter/backend/apps/common/form.py#L16) takes care of checking if the CPU is readOnly or not, whether to take the value from the request or the config, raising 400 etc.

So either you have a `cpu` value at that point or the backend has already raised a `400 BadRequest` exception. 

If that case, no `cpu` value in case of readOnly, is the only one for adding that check then we should be OK removing it. Is there something else that you think I missed?

Reply:
I just checked why I added this and it is for the following scenario. With `readOnly: false` and `limitFactor: "none"` I was encountering the following error:

```
2021-04-13 15:51:14,585 | kubeflow.kubeflow.crud_backend.errors.handlers | ERROR | Caught and unhandled Exception!
2021-04-13 15:51:14,585 | kubeflow.kubeflow.crud_backend.errors.handlers | ERROR | could not convert string to float: ''
Traceback (most recent call last):
  File "/home/david/.local/lib/python3.8/site-packages/flask/app.py", line 1950, in full_dispatch_request
    rv = self.dispatch_request()
  File "/home/david/.local/lib/python3.8/site-packages/flask/app.py", line 1936, in dispatch_request
    return self.view_functions[rule.endpoint](**req.view_args)
  File "/home/david/.local/lib/python3.8/site-packages/kubeflow/kubeflow/crud_backend/decorators.py", line 18, in wrapper
    return func(*args, **kwargs)
  File "/home/david/.local/lib/python3.8/site-packages/kubeflow/kubeflow/crud_backend/decorators.py", line 40, in runner
    return func(*args, **kwargs)
  File "/home/david/Documents/GitHub/kubeflow/components/crud-web-apps/jupyter/backend/apps/default/routes/post.py", line 30, in post_pvc
    form.set_notebook_cpu(notebook, body, defaults)
  File "/home/david/Documents/GitHub/kubeflow/components/crud-web-apps/jupyter/backend/apps/common/form.py", line 177, in set_notebook_cpu
    if float(cpu_limit) < float(cpu):
ValueError: could not convert string to float: ''
```

Reply:
Changing `if cpu_limit is not (None and ''):` to `if cpu_limit is not (None or ''):` would fix this problem, but causes the following error with `readOnly: true` and `limitFactor: "none"`:

```
2021-04-13 15:58:46,709 | kubeflow.kubeflow.crud_backend.errors.handlers | ERROR | Caught and unhandled Exception!
2021-04-13 15:58:46,709 | kubeflow.kubeflow.crud_backend.errors.handlers | ERROR | float() argument must be a string or a number, not 'NoneType'
Traceback (most recent call last):
  File "/home/david/.local/lib/python3.8/site-packages/flask/app.py", line 1950, in full_dispatch_request
    rv = self.dispatch_request()
  File "/home/david/.local/lib/python3.8/site-packages/flask/app.py", line 1936, in dispatch_request
    return self.view_functions[rule.endpoint](**req.view_args)
  File "/home/david/.local/lib/python3.8/site-packages/kubeflow/kubeflow/crud_backend/decorators.py", line 18, in wrapper
    return func(*args, **kwargs)
  File "/home/david/.local/lib/python3.8/site-packages/kubeflow/kubeflow/crud_backend/decorators.py", line 40, in runner
    return func(*args, **kwargs)
  File "/home/david/Documents/GitHub/kubeflow/components/crud-web-apps/jupyter/backend/apps/default/routes/post.py", line 30, in post_pvc
    form.set_notebook_cpu(notebook, body, defaults)
  File "/home/david/Documents/GitHub/kubeflow/components/crud-web-apps/jupyter/backend/apps/common/form.py", line 177, in set_notebook_cpu
    if float(cpu_limit) < float(cpu):
TypeError: float() argument must be a string or a number, not 'NoneType'
```

Reply:
These error are resolved by implementing the code from https://github.com/kubeflow/kubeflow/pull/5815#discussion_r612466379. I've just pushed a commit that removes this section.